File: Graphs/topoSort.cpp
#include<iostream>
#include<vector>
#include<stack>
using namespace std;

void dfs(int node, vector<int> adj[], vector<int> &vis, stack<int> &st){
    vis[node]=1;
    for(auto x : adj[node]){
        if(!vis[x]){
            dfs(x, adj, vis, st);
        }
    }
    st.push(node);
}

int main(){
    int n,m;
    cout<<"Enter the number of Nodes :: ";
    cin>>n;
    cout<<"Enter the number of Edges :: ";
    cin>>m;

    vector<int> adj[n+1];
    stack<int> st;
    for(int i=0;i<m;i++){
        int u,v;
        cin>>u>>v;
        adj[u].push_back(v);
    }
    vector<int> vis(n+1,0);

    for(int i=0;i<n;i++){
        if(!vis[i]){
            dfs(i, adj, vis, st);
        }
    }

    while(!st.empty()){
        cout<<st.top()<<" ";
        st.pop();
    }
    return 0;
}

File: Algos/Graphs/TopoSort.java
package Algos.Graphs;
import java.util.*;

public class TopoSort {
    static ArrayList<Integer>[] adj;
    static boolean[] visited;
    static Stack<Integer> st = new Stack<>();

    static void dfs(int node) {
    visited[node] = true;
    for (int x : adj[node]) {
      if (!visited[x]) {
        dfs(x);
      }
    }
    st.push(node);
  }
    
    /**
     * @param args
     */
    public static void main(String[] args) {
    Scanner sc = new Scanner(System.in);

    System.out.print("Enter the number of nodes: ");
    int n = sc.nextInt();

    System.out.print("Enter the number of edges: ");
    int m = sc.nextInt();

    adj = new ArrayList[n];
    for (int i = 0; i < n; i++) {
      adj[i] = new ArrayList<>();
    }

    for (int i = 0; i < m; i++) {
      int u = sc.nextInt();
      int v = sc.nextInt();
      adj[u].add(v);
    }

    visited = new boolean[n];
    for (int i = 0; i < n; i++) {
      if (!visited[i]) {
        dfs(i);
      }
    }

    while (!st.empty()) {
      System.out.print(st.pop() + " ");
    }
  }
    
}
